### DISCUSSION TITLE
Add way to control the "strictness" of the API errors

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#4368] [Feature Requests]
Currently, we surface the same generic 403 error for authorization errors _and_ 404s. This is done to make sure we don't accidentally expose the schema and data within. However, it's also a pretty bad developer experience, as there's no real way to know whether or not you're running into a permissions problem, or just have a typo in your route.

We should consider adding an environment variable that lets you control how much security to "trade in" for DX, we could start with:

| level | description |
|------|------------|
| strict | Generic 403s for everything |
| normal | 403s, but with descriptions\* |
| ?? friendly | 404s for non-existing items, authorization errors w/ descriptions |

\* Descriptions like "You don't have permission to access the foo collection"

(h/t @MiniDigger)

**Comment by MiniDigger** [comment]
Typo strict | Generic 403s for everything, should be 404

Edit: wait maybe not, I'll reread tomorrow ^^

**Reply by rijkvanzanten** [reply]
Haha no, that's correct 😉

**Reply by philleepflorence** [reply]
Extending your idea, I can imagine the [Error Management](https://en.wikipedia.org/wiki/Exception_handling) Levels to be:

- Strict
- Normal
- Friendly
- Debug: Includes error message and **file** or **component** that throws the error.

**Reply by RonaldKools** [reply]
And maybe also add an option for a 410 for archived items that have been online but no longer are...?

**Comment by lluishi93** [comment]
Some business requerimenets need more specific information given by the API in order to proceed with the logic. If the response is always 403 is impossible to know how to continue in certain cases (i.e. we try to upload some item from one environment to another, but that item does not exist in the target environment, how are we going to know?) It'd be useful to have this feature in order to receive more detailed errors

**Comment by battosai30** [comment]
I lost 30 minutes searching why I got 403 on specific ID even if I could get the entire collection. So clearly an error management could be a great improvement :)

**Comment by lluishi93** [comment]
Will this idea be implemented some day?

**Reply by rijkvanzanten** [reply]
If there's enough community interest, or somebody sponsors the implementation, sure thing! As of right now, it's not even [in the top-5 _pages_ of feature requests](https://github.com/directus/directus/discussions/categories/ideas?discussions_q=sort%3Atop+category%3AIdeas&page=1), so little hard to prioritize 🙂

**Reply by alishi973** [reply]
Is it possible to open MR for fixing this?

**Reply by MiniDigger** [reply]
I am sure the team appreciates anybody who comes forward to work on this!

**Reply by alishi973** [reply]
sounds interesting 😃

**Reply by rijkvanzanten** [reply]
> Is it possible to open MR for fixing this?

If there's a final spec that we all agree on! Right now, there's a vague idea of what this would mean in practice, but no fully formed spec of exactly what the configuration options are, or how it behaves in practice.

**Comment by rijkvanzanten** [comment]


**Reply by vanling** [reply]
![image](https://github.com/directus/directus/assets/50699657/50b428a3-c032-4bc7-a0f7-efe4736786f3)
hahaha  ❤️ 

ontopic: Would be very nice to see a more verbose log/error on permissions 👍

**Comment by slyk** [comment]
I'm new to Directus and lost in sum about a week getting same error on different stages of moving from other CMS to Directus. That is really strange and can lead someone to stop trying to understand errors and just think the api is broken and move to some alternative of Directus. I guess that is not good to community.

As example of other services, when new instance is installed, the messages level could be set to 'debug' because the instance is empty and nothing to protect. But just set some "WARN notification" or separate warning icon near the notifications icons that will tell that system is in debug/dev mode and to put in production you need to change settings.

That will help newcomers easy try functionality and debug SDK. Because when you on a "selecting DB" stage, you need to fast install and try different solutions to understand what is good for you.

**Comment by lluishi93** [comment]
Hi @rijkvanzanten  is there a way this can be done by custom coding without waiting for the feature itself to come out? For instance, using hooks or something like that? Thanks in advance.

**Reply by loprima-l** [reply]
I think we shall open a PR

**Reply by rijkvanzanten** [reply]
@lluishi93 Not really. Currently the `ForbiddenError` is thrown in places where it should be something else. To achieve this, we have to go over each use of `ForbiddenError` and replace it with something more meaningful. Once that's done, we have to update the `respond` middleware to return those as either forbidden or original based on the env var

**Reply by loprima-l** [reply]
@rijkvanzanten I've an idea on how to implement it in a consistent way I'll let you know it by a PR soon and if it pleases you and the members of the project I'll try to implement it (I never coded anything in the directus core before)

**Reply by rijkvanzanten** [reply]
@loprima-l Before we dive into coding, can we take a beat to finalize what the exact configuration options should be? On/off is an obvious one, but should we go the route of having different "levels" of strictness like in my original post?

**Reply by loprima-l** [reply]
Yeah, I don't have the energy to write down the full idea now (It's 11:30 pm in France) but to make it simple : adding a new category in the permissions will be an option that will work great to my opinion.

We can that way manage for each collection and also by the user type.

I plan on adding some warning if some of the configuration aren't expected.

And having an option in the config to enable the "full debug mode" (aka detailed logs and not just basics http code). 

Please let me know if you have any questions/doubts.

**Reply by rijkvanzanten** [reply]
Oh doing it in Permissions is a totally new angle I hadn't even considered! It would be really helpful if we could write up how that should exactly be configured and work, reach a conclusion, and then implement it 😄  Happy to chat through that here or on Discord, but not at 11:30PM 😉

**Reply by loprima-l** [reply]
I'm glad to help you getting through new points of vue !
I'm looking forward being useful in the directus project from a long time ago as Directus is in my 💜 

Maybe you should add me on discord (loprima is my ID) and we'll plan something there.

**Comment by rijkvanzanten** [comment]
@loprima-l and I have a working doc with notes / research here: https://directus.notion.site/Error-Friendliness-8ea55f15cdbf464dbea375f253005973

**Comment by richardragents** [comment]
There is progress in this process?

**Comment by antick** [comment]
Can't believe that nothing has been done about it yet. It's the most painful thing in directus.

**Reply by slyk** [reply]
well its a pain on the start. then you most of the times know what is wrong, and its mostly:
- permissions
- wrong id given

**Reply by antick** [reply]
But how can I possibly know which ID or permission is the problem without actually seeing the error? I have an admin role with full access and still getting the error while updating an image title.

**Comment by w0kyj** [comment]
Notes from Review:
- Why do we do it this way currently?
    - Security - prevent data model brute force discovery
- What would help improve this?
    - Prod vs Dev mode
    - Only allowed detailed errors to authenticated users
    - Allowed IPs
        - IP are ephemeral in cloud
    - 3 modes/levels
        - ???
- Implementation Ideas:
    - Error Management update to api
        - throw errors under the hood
        - add error translation handler to the last step of error processing
        - include a way to control which error translations to show via roles/permissions
    - Encrypted directus_errors collection
        - errors are stored in new collection
        - return the id
        - for details then query the collection with id and appropriate permissions
        - CAUTION: easy to overload/spam/table size/etc
        - Way to configure which errors get added -- could include deduplication, rate limits for spam/attack, ...
- Requirements
    - Must Have
        - Role/Policy level allow list type configuration (wildcard all for Admin)
        - Error Types defined -- include collection level flags
        - New operations to create/manage errors and return normalized error format
    - Should Have
        - ?
    - Could Have
        - Language translations for the messages
    - Will Not Have
        - ?
        
PLEASE FEEL FREE TO ADD/UPDATE :)

**Comment by jonaskohl** [comment]
This **MUST** be implemented **SOON** because **a lot of developer time gets wasted** debugging API access, which in turn costs companies using Directus **THOUSANDS OF DOLLARS**, thus **IMMEDIEATLY turning them away** from the otherwise invaluable tool that is Directus.

**Reply by rijkvanzanten** [reply]
You've **MADE** your **POINT** very **CLEAR**

**Reply by CoconutTheSlayer** [reply]
That is what I just did for the last 3 hours.... debugging 403, when I had a wrong payload

**Reply by Xchos** [reply]
You are not alone, @CoconutTheSlayer .. Same with me

**Comment by botafi** [comment]
+1 why is this not implemented yet?

**Comment by BorisKamp** [comment]
Another bump here, this is important Directus team!

I just spend 1.5 hour debugging my `403` which was very annoying as I was looking into permissions! I do not have a lot of knowledge about security but I dont understand how this can otherwise be a REALISTIC security threat? I mean, is it really dangerous to allow expose of schema by returning a 404 while in case to bum p against it you (as a user) already have explicit permission to read the entity OR the admins defined the entity as public readable. I do not get why this would be sensitive.

**Reply by rijkvanzanten** [reply]
You'd be surprised.. yeah.. At one point in time it did return 404s for non-existing collection/fields, and we did get slapped on the wrist with security vulns and customer penetration testing reports for exposing the data model. I can't get rid of it completely for that reason, hence my original idea being to have some sort of flag to toggle between options first.

**Comment by falmanna** [comment]
I think this is very important. It is the most time consuming problem devs are facing when building secure APIs with directus. 
I believe this can be simply tied to the logging system of directus. Basically, we don't have to return the details in the response, we just need them to be logged. So adding extra "Debug" level logs with details about what went wrong can be all we need!

**Reply by marekbrze** [reply]
Or a simple playground inside of the Directus to see the API dependencies between collections

**Comment by georgievdejan** [comment]
Are there any updates on this issue? Is there anything we can do to help?

We’re currently working on a project for a customer, but I can’t hand it over to them as long as this issue persists.

**Reply by rijkvanzanten** [reply]
No new updates yet. That being said, what would be incredibly helpful is to have a proper RFC specced version of this feature request. There's a lot of unknowns and assumptions in this thread around potential solves, so having one definitive spec to discuss, finalize, and implement would be very helpful indeed.

New feature requests already use (and enforce) [this format](https://github.com/directus/directus/blob/main/.github/DISCUSSION_TEMPLATE/draft-feature-requests.yml?plain=1).

**Reply by marekbrze** [reply]
I think the great minimal solution would be to receive permissions needed for selected query in the response when sending query using admin API key. Even as a list of requests that have been done to the other collections. I think the biggest pain point is the lack of knowledge which system collections should be also exposed in the permissions to successfully run the query. 

PS: unfortunately I don't have needed skills for coming up and implementing the solution for such a feature.

**Comment by alexvdvalk** [comment]
Perhaps the token type (static vs JWT) or flag on the user record or policy can determine which errors are returned.

**Comment by jclaveau** [comment]
### Workaround the DX issue
As everybody here, as soon as I developed with Directus I got this kind of misleading feedback.
I quite fast decided to fork Directus and my first patch was to add console.logs() to stop being blocked in my dev process.

### Cleanup
Presently in the process of upgrading Directus, I just wrote a draft adding a lot of `reasons` to ForbiddenErrors following the "new" [Errors implementation](https://github.com/directus/directus/pull/18797).

Here is my draft https://github.com/jclaveau/directus/pull/11

### ForbiddenError abstraction
Adding them, I realized ForbiddenErrors could be grouped in different use cases (based on the valuable vars I could pass to them):  
- [ ] `SystemCollectionAccessError` => directus_xxx
- [ ] `PermissionError` => checking accountability against collection / field / relation
- [ ] `AdminAccessRequiredError` => based on admin role
- [ ] `InvalidTokenError` => for the metrics controller
- [ ] `GenericForbiddenError` => Any forbidden error using specific vars like TFA... (non belonging to one of the other groups)
- [ ] `NotExposedMissingError` => All the forbidden errors that should normally be 404 (The topic of the current discussion)

The list may change based on this type https://github.com/jclaveau/directus/blob/v11.9.2-hhh-dev_permission-error-event/packages/errors/src/errors/forbidden.ts#L20


### Display logic
- These types may be an interesting criteria to display (or not) the reason to the user
    - `SystemCollectionAccessError` reason may only be displayed in dev mode
    - `PermissionError` is the most suitable for being displayed to any user
    - ...
- Also, my first goal was to trigger an emitFilter / emitAction, following the example of db errors.  Having this would make it easy to log the reasons while a generic `Forbidden` would be displayed to the user, letting the responsibility to the developer
- The last guard would be to use the env vars (dis)allow the display of the detailed reason


Please tell me if my approach seems feasible. I'd love to don't require a fork anymore!

Cheers !

**Comment by kazakevic** [comment]
Don't understand why collection name is treated as private data. E.g. if Directus is used directly with the frontend app w/o proxying requests by backend you will already see all collection names. At the end someone who need to bruteforce is not su dump so he will be stopped by 403 responses. Just need to have requests logs and secure authorazation part.

**Reply by rijkvanzanten** [reply]
> if Directus is used directly with the frontend app w/o proxying requests by backend you will already see all collection names

_if_ being the keyword there... We have a lot of different use cases, some of which are not internet / website facing. We've gotten pentest and vulnerability reports in the past for this exact reason. I'm personally with you that it feels a little overkill, but it is a real concern folks have
